### PR TITLE
[fix][broker] catch exception for brokerInterceptor

### LINKED ISSUE #19146: [Bug] Client receives duplicated response
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.

### Version

linux with pulsar  2.10.2

### Minimal reproduce step

1. implements a BrokerInterceptor and throws exceptions when interceptor `messageAcked`
2. start broker and enable brokerIntercetpor with the implemented by step1
3. send some messages by a producer
4. receive and ack messages by consumer
5. you will see logs in consumer like :
```shell
2023-01-04T19:33:49,654+0800 [pulsar-client-io-1-1] WARN  org.apache.pulsar.client.impl.ClientCnx - AckResponse has complete when receive response! requestId : 360167214930368447, consumerId : true
2023-01-04T19:33:49,654+0800 [pulsar-client-io-1-1] WARN  org.apache.pulsar.client.impl.ClientCnx - AckResponse has complete when receive response! requestId : 360167214930368448, consumerId : true
2023-01-04T19:33:49,654+0800 [pulsar-client-io-1-1] WARN  org.apache.pulsar.client.impl.ClientCnx - AckResponse has complete when receive response! requestId : 360167214930368449, consumerId : true
2023-01-04T19:33:49,654+0800 [pulsar-client-io-1-1] WARN  org.apache.pulsar.client.impl.ClientCnx - AckResponse has complete when receive response! requestId : 360167214930368450, consumerId : true

```

### What did you expect to see?

The client should not receive duplicated responses for the same request

### What did you see instead?

Duplicated ack response for one ack request.
```shell
2023-01-04T19:33:49,654+0800 [pulsar-client-io-1-1] WARN  org.apache.pulsar.client.impl.ClientCnx - AckResponse has complete when receive response! requestId : 360167214930368447, consumerId : true
2023-01-04T19:33:49,654+0800 [pulsar-client-io-1-1] WARN  org.apache.pulsar.client.impl.ClientCnx - AckResponse has complete when receive response! requestId : 360167214930368448, consumerId : true


```

### Anything else?

This is caused by the brokerInterceptor. 
```java
 consumer.messageAcked(ack).thenRun(() -> {
                if (hasRequestId) {
                    ctx.writeAndFlush(Commands.newAckResponse(
                            requestId, null, null, consumerId));
                }
               brokerInterceptor.messageAcked(this, consumer, copyOfAckForInterceptor);
            }).exceptionally(e -> {
                if (hasRequestId) {
                    ctx.writeAndFlush(Commands.newAckResponse(requestId,
                            BrokerServiceException.getClientErrorCode(e),
                            e.getMessage(), consumerId));
                }
                return null;
            });
```
After processing the `messageAck` request,
1.  broker will send a successful response to the consumer
2. broker will interceptor by brokerInterceptor

If the brokerInterceptor throws an exception,  this exception is caught by the broker too, and then sends a failed response to the consumer.
This failed response has the same requestId as the successful response, this means the consumer will receive two responses for one request.What's more, the failed response will be discarded directly, since the requestId has been removed by the successful response. So we can't get any exception or error logs from both the broker or the consumer.

There are some other places has the same problem, like:  handleSubscribe, buildProducerAndAddTopic.

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by aloyszhang**
Fixes #19146 

### Motivation

see #19146 

### Modifications

catch exception for brokerInterceptor

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

### Documentation
- [x] `doc-not-needed` 

### Matching PR in forked repository

PR in forked repository:

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java [hunk 1]
```diff
@@ -1157,7 +1157,11 @@ protected void handleSubscribe(final CommandSubscribe subscribe) {
                                         remoteAddress, topicName, subscriptionName);
                                 commandSender.sendSuccessResponse(requestId);
                                 if (brokerInterceptor != null) {
-                                    brokerInterceptor.consumerCreated(this, consumer, metadata);
+                                    try {
+                                        brokerInterceptor.consumerCreated(this, consumer, metadata);
```
Comment (congbobo184) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java:1161]: I am not sure if we need to send success before catching this exception. if it throws RuntimeException, maybe we should return an error first, not `sendSuccessResponse`.

**Review by BewareMyPower — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java (hunk 1 above)
Comment (BewareMyPower) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java:1161]: The exception thrown by an interceptor does not mean an error from broker, it's an error from 3rd party extensions. A similar case is that you can see `ConsumerInterceptors` , which catches all `Throwable`s from the user-configured interceptor.

Comment (aloyszhang) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java:1161]: The key problem now is the BrokerInterceptor exception is caught by the broker and then sends a failed response with a duplicated requestId.  So we can't get any information about the exception. 

For the exception processing, I agree with @BewareMyPower. It's better to catch`Interceptors` exceptions than throw them and return an error.

**Review by mattisonchao — APPROVED** [review]
LGTM

**Review by HQebupt — APPROVED** [review]
Good catch!

**Review by Jason918 — COMMENTED** [review]
I wonder if we should catch exceptions for all the methods in `BrokerInterceptor`.
Maybe we can add a wrapper class on top of BrokerInterceptor and handle all the exceptions in that class. 
Similar with `ConsumerInterceptors`

**Comment by aloyszhang** [comment]
> I wonder if we should catch exceptions for all the methods in BrokerInterceptor.

Sure, we can unify the exception processing for all methods.

**Review by eolivelli — CHANGES_REQUESTED** [review]
If the interceptor had a problem and it failed, maybe the system will be in an unconsistent state with corrupted data.
For instance if an interceptor has to mark all the messages with a special metadata or prevent a client from doing something and it failed, then basically the system behaves like the interceptor wasn't configured.

If the Interceptor throws an exception when it is not supposed to be allowed, then we must fail and the owner of the interceptor must fix the bug.

it is better to have a unavailable system then a system with corrupted data

**Comment by aloyszhang** [comment]
@eolivelli Thanks for your reminder.
BrokerInterceptor is a 3rd party extension and may have a variety of implementations and exceptions. If every interceptor exception may fail the broker, then, we have to take care of the exception handling wherever Interceptor is used.

And if the interceptor methods are not processed properly, the situation described in issue #19146 may occur(we got nothing for the exception, and the broker does not fail either)

> it is better to have a unavailable system then a system with corrupted data

I think we should make the best effort to keep the availability for broker.  
The developer should take care of the inconsistent state or corrupted data caused by BrokerInterceptor. And with the error log(printed at the catch block), we are able to troubleshoot these problems.

So, IMO, the exception of BrokerInterceptor should not affect the Broker process.

**Comment by Jason918** [comment]
> So, IMO, the exception of BrokerInterceptor should not affect the Broker process.

@aloyszhang I agree this should be the default behavior. But it would be better if `BrokerInterceptor` maintainer can decide if the exception should be thrown.

**Comment by BewareMyPower** [comment]
@eolivelli IMO, the interceptor should not be a fail-fast component. `Throwable` is caught for all other similar components currently. In addition to the `ConsumerInterceptor` example I've mentioned, the exception from `MessageListener` is also swallowed.

https://github.com/apache/pulsar/blob/badd69b8e012162feb28c86ca9de527cbfd1ac22/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBase.java#L1124-L1125

> But it would be better if BrokerInterceptor maintainer can decide if the exception should be thrown.

@Jason918 It could make things more complicated. It's unsafe to give an interceptor the power to terminate the broker.

**Comment by aloyszhang** [comment]
@eolivelli Need more feedback from you on whether BrokerInterceptor should throw an exception

**Comment by eolivelli** [comment]
> I think we should make the best effort to keep the availability for broker.

I think that here "availability" is not the right term or the most important thing.
If I have an interceptor that MUST, for instance, add a metadata header to every Message, and that is not happening due an internal error, we MUST NOT let the Message be stored.

If the system does not behave the way you expect then it is an unreliable tool and no body can use it in Production.

I am strongly opposed to this change with those blind catch clauses

**Comment by BewareMyPower** [comment]
> If I have an interceptor that MUST, for instance, add a metadata header to every Message, and that is not happening due an internal error, we MUST NOT let the Message be stored.

Actually the interceptor should not cause any side effect. But there is no way in Java to use a "const" parameter. If we want to allow the modifications, we should let the API return an error code or throw a checked exception to indicate if the modification succeeded.

If you allow the existing API to throw an exception and does not caught it, it might lead to an unexpected state at broker. Because the exception will be propagated and the upstream does not know how to handle the error. You can see the original issue (#19146) that the broker does not handle the exception well. It just sends the duplicated ACK responses.

Let's talk back to your scenario. Yeah in your case the message should not be stored. But what if the exception was just thrown by an implementation error of the interceptor? e.g. the interceptor try to parse the message value as an integer, but the value might not be an integer. In this case, the message should be stored and the exception that was not processed inside the interceptor should be swallowed by broker.

In short, to handle the case you mentioned, the best solution is to add a new API that throws a checked exception, not let the uncaught exception of the interceptor affect the broker.

**Comment by BewareMyPower** [comment]
I also found another reasonable explanation for why do we need to catch the exception in Kafka: https://github.com/apache/kafka/blob/2060b057b055b7e8cc62cbcf8aed710ca8bbe671/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerInterceptors.java#L63

> // do not propagate interceptor exception, log and continue calling other interceptors

If one interceptor throws an exception that is not swallowed, the rest interceptors would be skipped. But there is no way to handle the exception correctly (e.g. determine whether to store the message) except swallowing it.

**Review by eolivelli — CHANGES_REQUESTED** [review]
A couple of suggestions in order to move forward with this work, I don't want to block you but I also don't want to commit something that will make Pulsar not reliable.

1) Narrow down the change only to fix the issue that you are linking, here you are touching too many things
2) Add a test case that reproduces the problem and that validates that your change actually fixes it
3) If you want to tackle the problem in general about Interceptors and Errors, please start a broader discussion. But usually in a big community like ours it is better to work on smaller items at a time and reach consensus.

**Comment by BewareMyPower** [comment]
@eolivelli @aloyszhang Totally, I agree that we should narrow down the change only to avoid breaking changes. But further more, we might need to start a discussion about how the broker should handle the exception from the interceptor.

**Review by eolivelli — CHANGES_REQUESTED** [review]
I think that we are on our way now.

Next steps:
- handle InterruptedException (catch InterruptedException and then set Thread.currentThread.interrupt(), maybe we already have some utility method)
- add test cases

**Comment by Jason918** [comment]
> @Jason918 It could make things more complicated. It's unsafe to give an interceptor the power to terminate the broker.

@BewareMyPower This can not be simply categorized as `complicated`. It's not sufficient enough to decide we don't do something.
I agree we need more discussion on this topic. From what I see, the interceptor author should have more insight on whether it should throw the errors, like the case Enrico mentioned. And they are the core developers of Pulsar, they should know Pulsar works and what happens if the interceptor throws the error.

**Comment by BewareMyPower** [comment]
@Jason918 It's not related to whether the interceptor developers are core developers of Pulsar. It's related to the fact that there is no clear definition in Pulsar about how the broker handles an exception from the interceptor. I said "complicated" because they have to look deeper into the code when they found the interceptor threw an exception and caused an undefined behavior like this PR and its related issue.

**Comment by aloyszhang** [comment]
PTAL @eolivelli

**Comment by aloyszhang** [comment]
cc @eolivelli

**Review by eolivelli — APPROVED** [review]
LGTM

The places in which we are catching any errors are "after" the events happened, so any interceptor could not influence the execution of the main code, so it is fine to me to ignore any errors.
